### PR TITLE
[Auto Loading] Autoload extension settings

### LINKED ISSUE #11404: Error setting config `TimeZone` with Python client on `connect`
### What happens?

Trying to set `TimeZone` with Python client with `connect()` by passing a config dictionary causes `InvalidInputException`;

AFAIK from docs ICU extension is loaded by default on Python and reference to configuration options in Python docs contains `TimeZone` as a configuration option;
https://duckdb.org/docs/api/python/overview.html#configuration
https://duckdb.org/docs/configuration/overview#configuration-reference

```python
import duckdb
conn.sql("select * from duckdb_extensions() where extension_name = 'icu'").show()
```

```
┌────────────────┬─────────┬───────────┬──────────────┬──────────────────────────────────────────────────────────────────┬───────────┐
│ extension_name │ loaded  │ installed │ install_path │                           description                            │  aliases  │
│    varchar     │ boolean │  boolean  │   varchar    │                             varchar                              │ varchar[] │
├────────────────┼─────────┼───────────┼──────────────┼──────────────────────────────────────────────────────────────────┼───────────┤
│ icu            │ true    │ true      │ (BUILT-IN)   │ Adds support for time zones and collations using the ICU library │ []        │
└────────────────┴─────────┴───────────┴──────────────┴──────────────────────────────────────────────────────────────────┴───────────┘
```

Setting it with `SET ...` directive seems to work fine;
```python
import duckdb

conn = duckdb.connect()
conn.sql("select current_timestamp").show()
conn.sql("set TimeZone = 'UTC'")
conn.sql("select current_timestamp").show()
```

```
┌────────────────────────────┐
│     current_timestamp      │
│  timestamp with time zone  │
├────────────────────────────┤
│ 2024-03-28 16:41:07.292+03 │
└────────────────────────────┘

┌────────────────────────────┐
│     current_timestamp      │
│  timestamp with time zone  │
├────────────────────────────┤
│ 2024-03-28 13:41:07.293+00 │
└────────────────────────────┘
```

### To Reproduce

```python
import duckdb
conn = duckdb.connect(config={"TimeZone" : "UTC"})
```

```
Traceback (most recent call last):
  File "repro.py", line 3, in <module>
    conn = duckdb.connect(config={"TimeZone" : "UTC"})
duckdb.duckdb.InvalidInputException: Invalid Input Error: Unrecognized configuration property "TimeZone"
```

### OS:

macOS

### DuckDB Version:

0.10.1

### DuckDB Client:

Python

### Full Name:

Selman Kayrancioglu

### Affiliation:

Peak

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a release build (and could not test with a nightly build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #11404 

`TimeZone` is an `extension_setting` added by the ICU extension, so it's not available until the database has started.
When it's added to the config, it gets put into the `unrecognized_options` map.

After the database has initialized, we now look through the `unrecognized_options` to see if it contains settings we recognize as being part of a known extension and attempt to autoload them if possible.

### Running the added c++ test

Configuration:
`extension/extension_config_local.cmake`:
```
duckdb_extension_load(icu DONT_LINK)
```

Build:
`ENABLE_EXTENSION_AUTOLOADING=1 ENABLE_EXTENSION_AUTOINSTALL=1 make debug`

Run:
`LOCAL_EXTENSION_REPO=./build/debug/repository ./build/debug/test/unittest "Test autoload of extension settings"`

### CODE REVIEW

**Comment by Giorgi** [comment]
If the extensions are now automatically loaded if the settings are included in `duckdb_config` can we have the extension settings included in the [Configuration API](https://duckdb.org/docs/api/c/config.html) ?

Currently, it's impossible to determine whether a given configuration option is valid or not because the configuration API returns only core configs

File: src/main/database.cpp [hunk 1]
```diff
@@ -246,9 +292,8 @@ void DatabaseInstance::Initialize(const char *database_path, DBConfig *user_conf
 		ExtensionHelper::LoadExternalExtension(*this, *config.file_system, config.options.database_type);
 	}
 
-	if (!config.options.unrecognized_options.empty()) {
-		ThrowExtensionSetUnrecognizedOptions(config.options.unrecognized_options);
-	}
+	LoadExtensionSettings();
+	ThrowExtensionSetUnrecognizedOptions(config.options.unrecognized_options);
```
[the remark this answers is not in the document]
Comment (carlopi) [reply src/main/database.cpp:296]: Actually, I would move ThrowExtensionSetUnrecognizedOptions inside LoadExtensionSettings, and have that called `LoadExtensionsSettingsOrThrow`

**Review by carlopi — COMMENTED** [review]
I am not super comfortable with this change, since it expand quite a bit the surface of when it become possible to load extensions.

I would probably restrict any loading that happens here to core extensions, but we don't have a mechanism for that yet. In particular since extensions might bring in other extensions on Load.

This is not possible right now, due to the fact that CanAutoload works as an allow-list of sort, so maybe it's fine after all, but in that case I will add a // FIXME in the code with a multi line comment explaining that this is possibly problematic and needs to be revisited. I would consider a comment also in CanAutoload, since that function now acts as allow-list.

My preferred way out would be probably to have an option to specify a list of extensions to be AutoLoaded during Initialization. And those can be checked / vetted more explicitly by an appropriate function that should temporarily forbid community_extensions, do the load, and re-enable community-extension to their default.
So that if you specify "timezone", you also need to specify "icu" in the list of extensions to be loaded.

**Comment by Tishj** [comment]
> My preferred way out would be probably to have an option to specify a list of extensions to be AutoLoaded during Initialization.

This is enabled through `EXTENSION_SETTINGS` in `extension_entries.hpp`, we're not loading/installing arbitrary extensions from the web.

Right before LoadExtensionSettings() we already potentially load and extension based on the `database_type`. I don't really get why doing this additional autoload based on known extension settings that are explicitly provided by the user is problematic, e.g:
```py
 con = duckdb.connect(config={"TimeZone": "UTC"})
```

**Comment by carlopi** [comment]
I was taking another look, SetICUTimeZone is NOT called, and this is yet another difference: extension can execute code in callbacks.

To check this, try:
```
SET TimeZone='America/Argh';
```
in the CLI it throws:
```
Not implemented Error: Unknown TimeZone setting
```
while in the "Test autoload of extension settings" will work (and be set to a nonsensical value).

**Comment by Tishj** [comment]
```c++
void DBConfig::SetOptionByName(const string &name, const Value &value) {
	auto option = DBConfig::GetOptionByName(name);
	if (option) {
		SetOption(*option, value);
		return;
	}

	auto param = extension_parameters.find(name);
	if (param != extension_parameters.end()) {
		Value target_value = value.DefaultCastAs(param->second.type);
		SetOption(name, std::move(target_value));
	} else {
		options.unrecognized_options[name] = value;
	}
}
```

This is broken, it's completely ignoring the `set_option` callback of the ExtensionOption.
This is likely because of the ClientContext & in the `set_option` that we can't provide here

File: src/main/database.cpp [hunk 2]
```diff
@@ -194,13 +194,50 @@ void DatabaseInstance::CreateMainDatabase() {
...
+			auto extension_name = ExtensionHelper::FindExtensionInEntries(name, EXTENSION_SETTINGS);
+			if (extension_name.empty()) {
+				continue;
+			}
+			if (!ExtensionHelper::TryAutoLoadExtension(*this, extension_name)) {
+				throw InvalidInputException(
+				    "To set the %s setting, the %s extension needs to be loaded. But it could not be autoloaded.", name,
+				    extension_name);
+			}
+			config.SetOptionByName(name, value);
```
Comment (Mytherin) [inline src/main/database.cpp:230]: I think you have to use a different method here to make sure the set option callback is called correctly

Comment (Mytherin) [reply src/main/database.cpp:230]: I think we can just make a connection here - let's give that a shot first before we try to modify the callback

**Review by carlopi — APPROVED** [review]

**Comment by carlopi** [comment]
I think this is ready to be considered for merge.

I had tried the problematic test locally, fails as expected, looks good to me.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**